client.go

```diff
@@ -326,7 +326,7 @@ func (c *client) recv(l *net.UDPConn, msgCh chan *dns.Msg) {
 		return
 	}
 	buf := make([]byte, 65536)
-	for !c.closed {
+	for closed(c) {
 		n, err := l.Read(buf)
```

Comment:
you're still missing the case when Read() unblocks after a connection Close(), Read() will still return error when we just don't care about that error.